### PR TITLE
Deduplicate various steps in CI files

### PR DESCRIPTION
**Pull request by krzysz00**
- Add most of the flags we use in CI to the CMakeLists.txt

 - Introduce a BUILD_FAT_LIBMLIRMIOPEN to control building the fat MIOpen interface library and avoid building it when it's not needed

 - Move the installation of said library to a separate component, so  now we don't need to run our build tools with root permissions

 - Replace many shell scripts that call cmake and a build tool with calls to the cmake plugin

- Similarly, move PATH changes into the Jenkins configuration

Many of these CI files still contain similar steps (like various
compile/test combinations). Once I have a better sense of what tests
are being run and where, I will be able to combine these files into a
something less redundant and include, for example, conditions for
excluding longer tests on PR-triggered CI.

### CODE REVIEW

**Comment by krzysz00** [comment]
I'll want to hold off on merging this until #223 lands, since that'll affect the changes I'm making and change the flags I'm supposed to set. However, I'll still take review

File: mlir/utils/jenkins/Jenkinsfile.downstream
Comment (yiqian1) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: Also need -DMLIR_MIOPEN_DRIVER_PR_E2E_TEST_ENABLED=1 here.

Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: `git merge` messed that up, fixed, thanks for catching it

File: mlir/tools/mlir-miopen-lib/CMakeLists.txt
Comment (jerryyin) [inline mlir/tools/mlir-miopen-lib/CMakeLists.txt:83]: [Question] Just want to confirm that there's no change of behavior here. Building the target still install the library?

File: mlir/utils/jenkins/Jenkinsfile.downstream
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.downstream:10]: I'm not sure this is required to build/check-mlir. I think this is required on `Static Test` stage. If that's the case, could you add a comment to be explicit on which step need this environment?

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.miopen-test]: This copies the build step package to test stage? Nice!

**Review by jerryyin — COMMENTED** [review]
Do you see the possibility to further consolidate everything into 2 jenkinsfile
 - 1 nightly jenkins file
 - 1 PR triggered jenkins file

File: mlir/utils/jenkins/Jenkinsfile.downstream
Comment (yiqian1) [inline mlir/utils/jenkins/Jenkinsfile.downstream]: Should this be --target check-mlir?
I'm sure the tools were built but tests were not run. Please check the console outputs.

File: mlir/tools/mlir-miopen-lib/CMakeLists.txt
Comment (krzysz00) [reply mlir/tools/mlir-miopen-lib/CMakeLists.txt:83]: There is a change of behavior. Building the target now does *not* install the library, hence the calls in the Jenkins file to `cmake --install`

File: mlir/utils/jenkins/Jenkinsfile.downstream
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.downstream:10]: I figured  that, since it's used all over the place, this should be placed on the top to avoid duplication (especially since it won't break anything)

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test]: Yep.

File: mlir/utils/jenkins/Jenkinsfile.downstream
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.downstream]: Could be, and I might remove the withCmake since it really isn't needed

**Comment by krzysz00** [comment]
Re consilidaiton: I'm currently writing a Jenkinsfile that merges everything into one file that's parametrized by public/private and nightly/PR

File: mlir/tools/mlir-miopen-lib/CMakeLists.txt
Comment (jerryyin) [reply mlir/tools/mlir-miopen-lib/CMakeLists.txt:83]: We synced offline, the new build install sequence is:

```bash
cmake -G "Unix Makefiles" .. -DCMAKE_BUILD_TYPE=Release -DBUILD_FAT_LIBMLIRMIOPEN=1
make -j libMLIRMIOpen
cmake --install . --component libMLIRMIOpen
```

File: mlir/test/Examples/standalone/test.toy
Comment (jerryyin) [inline mlir/test/Examples/standalone/test.toy:2]: We don't care about this directory. However, do you see the need of filing a ticket against ourselves in fixing the cmake version?

File: mlir/utils/jenkins/Jenkinsfile.upstream
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.upstream:24]: Why does this `sourceDir` matter? Why is it not `mlir`?

**Review by jerryyin — COMMENTED** [review]
LGTM. You have my approval. 

@yiqian1 Due to the magnitude of changes in this PR, please help take a second look. You are more familiar with each individual jenkinsfile. Approve and merge if you are also okay with it.

File: mlir/test/Examples/standalone/test.toy
Comment (yiqian1) [reply mlir/test/Examples/standalone/test.toy:2]: This test used to pass. What caused it to fail?

Comment (krzysz00) [reply mlir/test/Examples/standalone/test.toy:2]: It's not clear

**Comment by krzysz00** [comment]
Good news, I found a workaround for the MIOpen test, see http://10.236.106.97:8080/blue/organizations/jenkins/MLLIR-MIopen-Resnet50-refactor-dev/detail/MLLIR-MIopen-Resnet50-refactor-dev/35/pipeline/526 (turns out the issue is that we can't have a docker agent definition nested inside another docker agent definition)

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (jerryyin) [inline mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: 'rocm/mlir:rocm4.1-latest' should be equivalent with 'rocm/mlir:latest', since ROCm 4.1 is what we got.

**Comment by jerryyin** [comment]
> I found a workaround for the MIOpen test

Good job!

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: For unknown reasons, rocm4.1-latest didn't have `parallel` in it, maybe it's because we ROCm 4.2 has been released so the update didn't get given that tag?

**Comment by yiqian1** [comment]
LGTM. Just wonder what caused the standalone test to fail.

File: mlir/utils/jenkins/Jenkinsfile.upstream
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.upstream:24]: `sourceDir` here is the directory containing the CMakeLists.txt, and, in the previous Jenkinsfile, we built from `llvm` (and that's what our current setup also does - it builds `llvm`, which then goes on to include `mlir` under `tools/`)

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: > maybe it's because we ROCm 4.2 has been released

As long as we haven't bumped to ROCm 4.2 in the dockerfile, it won't happen automatically.

I took a look a second ago and see `rocm/mlir:latest` and `rocm4.1-latest` having the exact same commit hash. Can you try to pull again, I suspect you have a out-dated image.

File: mlir/test/Examples/standalone/test.toy
Comment (jerryyin) [reply mlir/test/Examples/standalone/test.toy:2]: @krzysz00 I remember you mentioned that it is related with cmake. What's the last passing cmake version?

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: I'm not sure how to get Jenkins to re-pull

File: mlir/test/Examples/standalone/test.toy
Comment (krzysz00) [reply mlir/test/Examples/standalone/test.toy:2]: I can try and take a look at it

File: mlir/utils/jenkins/Jenkinsfile.upstream
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.upstream:24]: I'm not convinced this does anything. Taking a look back, I think that `../llvm` is added by mistake and you can build with `..` all the same - We do also have a `CMakeLists.txt` in the root directory.

Since we are not using this jenkinsfile anyways, I'm fine with you keep this as-is.

Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.upstream:24]: If you look at the context, this isn't our build - this is a build of LLVM from https://github.com/llvm/llvm-project - which doesn't have a CMakeLists.txt in the root.

However, if I misunderstood the setup and the idea is to test our code but with upstream's files, then I can change it.

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.upstream:24]: > this is a build of LLVM from

You get the idea right. Though, I still remember when I used to compile the tip of upstream llvm and don't need to do `../llvm` in the configuration step. 

I think this is an unimportant implementation detail that can always get verified easily when we start to build upstream tip again. I'm marking this as resolved.

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: I did some googling and find [this image](https://i.stack.imgur.com/ZP0vN.png), [here](https://stackoverflow.com/questions/46327298/jenkins-always-pull-docker-image-from-a-declarative-pipeline-agent-definition). Could you verify this? 

`alwaysPull true`

I think the idea is to force jenkins pull the latest docker instead of using the local out-of-dated cached image.

Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: We may not want to *always* have that on, but it's probably a good idea since we've recently made changes to our Docker images

File: mlir/tools/mlir-miopen-lib/CMakeLists.txt
Comment (krzysz00) [reply mlir/tools/mlir-miopen-lib/CMakeLists.txt:83]: You don't necessarily need to call for Makefiles - you could also, say, call for Ninja, or just leave `-G` off and let it default to Makefiles

File: mlir/utils/jenkins/Jenkinsfile.miopen-test
Comment (krzysz00) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: `alwaysPull true` does what we want - it always pulls, but `pull` itself won't redownload if nothing's changed

We may just want to check that we're not building new images too often.

Comment (jerryyin) [reply mlir/utils/jenkins/Jenkinsfile.miopen-test:68]: Thanks for confirming. 

> pull itself won't redownload if nothing's changed

That's pretty reasonable behavior.

> We may just want to check that we're not building new images too often.

Agreed, but I think we are already doing the best we can. Besides, the overhead of downloading a fresh image isn't that much so I won't worry about it now.

File: mlir/tools/mlir-miopen-lib/CMakeLists.txt
Comment (jerryyin) [reply mlir/tools/mlir-miopen-lib/CMakeLists.txt:83]: > or just leave -G off and let it default to Makefiles

Thanks for the suggestion

**Review by jerryyin — APPROVED** [review]
Solid work! Thanks

**Comment by whchung** [comment]
@jerryyin / @krzysz00 This PR needs to be reverted as it breaks `check-mlir`.

**Comment by whchung** [comment]
Filed #233 to revert this PR.

File: mlir/test/Examples/standalone/test.toy
Comment (whchung) [inline mlir/test/Examples/standalone/test.toy:2]: Why is this needed? It's breaking `check-mlir`

Comment (whchung) [reply mlir/test/Examples/standalone/test.toy:2]: Need to get to the bottom of this.

**Pull request merged**
